### ISSUE TITLE
How to run a custom python script every time the user edits their code? [eg. Linter extension]

### ISSUE THREAD

**Issue by mkery** [jupyterlab/jupyterlab#4610]
Hi! I am writing a Jupyterlab extension which, just like a linter, needs to run a quick little program analysis to detect properties of the user's code each time they make an edit.

1. My extension retrieves the user's active python kernel
2. I've got listeners `editor.model.value.changed` and `editor.model.selections.changed` so my extension knows when a code cell's content has changed
3. I've got a python script `analysis.py`

How do I get the kernel to run my` analysis.py`?

@jasongrout pointed my to the completer and inspector packages, which seem to have their own custom formatting to send a shell message through the kernel eg. `ICompleteRequest` and `IInspectRequest`. On the other end, in IPython,  eg. [completerlib.py](https://github.com/ipython/ipython/blob/bd3142f5f1b3b58ed917bda609d7db788d0591fb/IPython/core/completerlib.py) the completer and inspect messages have their built in python implementations.

I want to create my own message request type, `IAnalysisRequest`. How can I run arbitrary python code read from `analysis.py` that is not built into IPython? I do not want to use `kernel.execute` because I do not want to pollute the user's namespace with my program analysis functions. **Help!**

**Comment by mkery** [comment]
Adding another sub-problem: (maybe due to security/web-pack) I cannot seem to use `fs` or `fs-extra` in my extension to load in my `analysis.py`. The extension will not build but not load in Jupyter if I attempt to use either of those. As a (pretty awful) workaround, I'm temporarily just putting the contents of `analysis.py` in a string in my Typescript per the examples in packages/services/examples. Is there a better way to pass a python script file to the kernel?

**Comment by BoPeng** [comment]
This sounds to me like a server + lab extension combo where a server extension handles a custom command and a lab extension passes text to the server extension and displays results. It would be much easier and perhaps faster to rewrite your linter in JS so that you can do everything in the frontend though.

**Comment by BoPeng** [comment]
Also, in terms of how to send the linter message to the frontend and how to display the message, a `transient_display_data` message is [being discussed](https://github.com/jupyter/jupyter_client/issues/376) and a UI has been [proposed](https://github.com/jupyterlab/jupyterlab/issues/4550).

**Comment by jasongrout** [comment]
> This sounds to me like a server + lab extension combo where a server extension handles a custom command and a lab extension passes text to the server extension and displays results.

Good point. @mkery - do you need your analysis code to be executed in the context of the current kernel (i.e., inspecting current kernel state, or doing something specific to that exact kernel and the variables defined therein), or do you just need some generic python code to be executed, and the kernel is just a convenient place to execute the python code? If you just need some generic python code to be executed (and it doesn't matter which exact kernel you are in), then I agree that a server extension is probably a better way to go.
